src/pynwb/__init__.py

```diff
@@ -190,7 +190,7 @@ class NWBHDF5IO(HDF5IO):
 
     @docval({'name': 'path', 'type': str, 'doc': 'the path to the HDF5 file'},
             {'name': 'mode', 'type': str,
-             'doc': 'the mode to open the HDF5 file with, one of ("w", "r", "r+", "a", "w-")', 'default': 'a'},
+             'doc': 'the mode to open the HDF5 file with, one of ("w", "r", "r+", "a", "w-")', 'default': ''},
```

Comment:
wouldn't it be better to have no default value?

Reply:
You mean no like in `None`?

Reply:
```suggestion
             'doc': 'the mode to open the HDF5 file with, one of ("w", "r", "r+", "a", "w-")'},
```

Reply:
a `'default'` term is not required. This would make it a non-optional argument, which I believe is what you want

Reply:
@bendichter Thanks for the clarifcation. Yes I want that!

Reply:
If I leave out the default key I'll get something like

```
    @docval({'name': 'obj', 'type': (Container, Builder), 'doc': 'the object to get the ObjectMapper for'},
            returns='the ObjectMapper to use for mapping the given object', rtype='ObjectMapper')
    def get_map(self, **kwargs):
        """ Return the ObjectMapper object that should be used for the given container """
        obj = getargs('obj', kwargs)
        # get the container class, and namespace/data_type
        if isinstance(obj, Container):
            container_cls = obj.__class__
            namespace, data_type = self.get_container_ns_dt(obj)
            if namespace is None:
>               raise ValueError("class %s is not mapped to a data_type" % container_cls)
E               ValueError: class <class 'pynwb.file.NWBFile'> is not mapped to a data_type
```

I'll stick with None for now then.

Reply:
I'm confused with this error. Our convention for required args is to leave out the default key, so I'd like to make that work if possible. Let me check out your branch and see if I can make it work.

Reply:
OK I can recreate this error but I still haven't figured out what's going on here. Let me give this a bit more time, and as a plan B we can make the default arg `None`

Reply:
I have not looked at how docval handles this in detail, so my memory might be wrong, but here is what I think is going on.

I believe @bendichter is correct. Setting ``default=None`` I believe implies that None is a keyword argument with a default value of None. Since you didn't change how mode is being handled in the constructor, I think this means in this case that that self.mode in HDF5IO is being set to None and I think in turn h5py will end up choosing a mode, so effectively you still have an implicit default (i.e., whatever h5py does when you do h5py.File(..., mode=None)). I think this is also why the galleries are still passing tests, because there are quite a few instances in our tutorial galleries that do not set the mode, and all of those should fail tests if mode is a required argument. Note, when supplying no default arg in docval, then that means mode becomes a positional argument so one reason for the error you see could be that you inadvertently changed the order in which arguments are expected (which might explain the error you see).

Reply:
@t-b if I remember correctly, docval allows you to still set positional arguments via keyword assignemnt, i.e,. you should still be able to do ``mode='r'`` even when not specifying a default. In general, I think if ``mode`` is a required argument then we should do this in docval by leaving out the default key. The reason is, that in this way it shows up correctly as a required argument in the documentation and help, whereas if you set ``default=None`` then it looks like an optional argument in the documentation and then raise an error if a users actually uses the default.

Reply:
@oruebel Makes sense. Done.